components/camera/videosource/webcam.go

```diff
@@ -215,108 +125,69 @@ func makeConstraints(conf *WebcamConfig, logger logging.Logger) mediadevices.Med
+// as well as the path to the driver.
+func findReaderAndDriver(
+	conf *WebcamConfig,
 	path string,
-	fromLabel bool,
-	constraints mediadevices.MediaStreamConstraints,
 	logger logging.Logger,
-) (gostream.MediaSource[image.Image], error) {
-	if !fromLabel {
+) (video.Reader, driverutils.Driver, string, error) {
+	mediadevicescamera.Initialize()
+	constraints := makeConstraints(conf, logger)
+
+	// Handle specific path
+	if path != "" {
 		resolvedPath, err := filepath.EvalSymlinks(path)
 		if err == nil {
 			path = resolvedPath
 		}
-	}
-	return gostream.GetNamedVideoSource(filepath.Base(path), constraints, logger)
-}
-
-// tryWebcamOpen is a helper for finding and making the video source that uses getNamedVideoSource to try and find
-// a video device (gostream.MediaSource). If successful, it will wrap that MediaSource in a VideoSource.
-func tryWebcamOpen(
-	ctx context.Context,
-	conf *WebcamConfig,
-	path string,
-	fromLabel bool,
-	constraints mediadevices.MediaStreamConstraints,
-	logger logging.Logger,
-) (gostream.VideoSource, error) {
-	source, err := getNamedVideoSource(path, fromLabel, constraints, logger)
-	if err != nil {
-		return nil, err
-	}
-
-	if conf.Width != 0 && conf.Height != 0 {
-		img, release, err := gostream.ReadMedia(ctx, source)
-		if release != nil {
-			defer release()
-		}
+		reader, driver, err := getReaderAndDriver(filepath.Base(path), constraints, logger)
 		if err != nil {
-			return nil, err
-		}
-		if img.Bounds().Dx() != conf.Width || img.Bounds().Dy() != conf.Height {
-			return nil, errors.Errorf("requested width and height (%dx%d) are not available for this webcam"+
-				" (closest driver found by gostream supports resolution %dx%d)",
-				conf.Width, conf.Height, img.Bounds().Dx(), img.Bounds().Dy())
+			return nil, nil, "", err
 		}
-	}
-	return source, nil
-}
-
-// getPathFromVideoSource is a helper function for finding and making the video source that
-// returns the path derived from the underlying driver via MediaSource or an empty string if a path is not found.
-func getPathFromVideoSource(src gostream.VideoSource, logger logging.Logger) string {
-	labels, err := gostream.LabelsFromMediaSource[image.Image, prop.Video](src)
-	if err != nil || len(labels) == 0 {
-		logger.Errorw("could not get labels from media source", "error", err)
-		return ""
-	}
 
-	return labels[0] // path is always the first element
-}
-
-// findAndMakeVideoSource finds a video device and returns a video source with that video device as the source.
-func findAndMakeVideoSource(
-	ctx context.Context,
-	conf *WebcamConfig,
-	path string,
-	logger logging.Logger,
-) (gostream.VideoSource, string, error) {
-	mediadevicescamera.Initialize()
-	constraints := makeConstraints(conf, logger)
-	if path != "" {
-		cam, err := tryWebcamOpen(ctx, conf, path, false, constraints, logger)
-		if err != nil {
-			return nil, "", errors.Wrap(err, "cannot open webcam")
+		if conf.Width != 0 && conf.Height != 0 {
```

Comment:
Should we be checking conf width and height here? Could the conf specify 0s but the driver that is matched has a valid size?

Reply:
you're so right let me fix it to check regardless of config